run_linux.go

```diff
@@ -19,17 +19,260 @@ package main
 import (
 	"context"
 	"fmt"
-	"strings"
-
-	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/cio"
 	"github.com/containerd/containerd/oci"
 	"github.com/containerd/containerd/pkg/cap"
+	gocni "github.com/containerd/go-cni"
 	"github.com/containerd/nerdctl/pkg/defaults"
+	"github.com/containerd/nerdctl/pkg/dnsutil"
+	"github.com/containerd/nerdctl/pkg/dnsutil/hostsstore"
+	"github.com/containerd/nerdctl/pkg/logging"
+	"github.com/containerd/nerdctl/pkg/netutil"
+	"github.com/containerd/nerdctl/pkg/portutil"
+	"github.com/containerd/nerdctl/pkg/strutil"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
 	"github.com/urfave/cli/v2"
+	"net/url"
+	"os"
+
+	"path/filepath"
+	"strings"
 )
 
-func WithoutRunMount() func(ctx context.Context, client oci.Client, c *containers.Container, s *oci.Spec) error {
-	return oci.WithoutRunMount
+func NewContainer(ctx context.Context, clicontext *cli.Context, client *containerd.Client, dataStore, id string) (containerd.Container, error) {
+	var (
+		opts  []oci.SpecOpts
+		cOpts []containerd.NewContainerOpts
+	)
+
+	opts = append(opts,
+		oci.WithDefaultSpec(),
+		oci.WithDefaultUnixDevices,
+		oci.WithMounts([]specs.Mount{
+			{Type: "cgroup", Source: "cgroup", Destination: "/sys/fs/cgroup", Options: []string{"ro", "nosuid", "noexec", "nodev"}},
+		}),
+		oci.WithoutRunMount, // unmount default tmpfs on "/run": https://github.com/containerd/nerdctl/issues/157
+	)
+
+	rootfsOpts, rootfsCOpts, ensuredImage, err := generateRootfsOpts(ctx, client, clicontext, id)
+	if err != nil {
+		return nil, err
+	}
+	opts = append(opts, rootfsOpts...)
+	cOpts = append(cOpts, rootfsCOpts...)
+
+	if wd := clicontext.String("workdir"); wd != "" {
+		opts = append(opts, oci.WithProcessCwd(wd))
+	}
```

Comment:
Can be moved to run.go

Reply:
The `opts` variable is local to this function.  I could split out some of the common parameters and pass in opts to `NewContainer` but when I started with that it felt like the similarities were few and just enough differences it felt awkward keeping track.  If you want I can do it again as separate commit to see how it looks

Reply:
Yes, can be another PR